include/unifex/any_sender_of.hpp

```diff
@@ -0,0 +1,110 @@
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include <unifex/any_unique.hpp>
+#include <unifex/sender_concepts.hpp>
+
+#include <unifex/detail/prologue.hpp>
+
+namespace unifex {
+
+template <typename... Values>
+using any_receiver_of =
+    any_unique_t<
+        overload<void(this_&&, Values...)>(set_value),
+        overload<void(this_&&, std::exception_ptr) noexcept>(set_error),
+        overload<void(this_&&) noexcept>(set_done)>;
+
+using any_operation_state =
+    any_unique_t<
+        overload<void(this_&) noexcept>(start)>;
+
+namespace _any {
+
+// For in-place constructing non-movable operation states.
+// Relies on C++17's guaranteed copy elision.
+template <typename Fun>
+struct _rvo {
+  Fun fun_;
+  operator callable_result_t<Fun>() && {
+    return ((Fun&&) fun_)();
+  }
+};
+
+template <typename Fun>
+_rvo(Fun) -> _rvo<Fun>;
+
+template <typename... Values>
+struct _connect_fn {
+  using type_erased_signature_t =
+      any_operation_state(this_&&, any_receiver_of<Values...>&&);
+
+  template(typename Sender)
+      (requires sender_to<Sender, any_receiver_of<Values...>>)
+  friend any_operation_state
+  tag_invoke(_connect_fn, Sender&& s, any_receiver_of<Values...>&& r) {
+    return any_operation_state{
+      std::in_place_type<connect_result_t<Sender, any_receiver_of<Values...>>>,
+      _rvo{[&]() { return connect((Sender&&) s, std::move(r)); }}
+    };
+  }
+
+  template(typename Self)
+      (requires tag_invocable<_connect_fn, Self, any_receiver_of<Values...>>)
+  any_operation_state operator()(Self&& s, any_receiver_of<Values...>&& r) const {
+    return tag_invoke(*this, (Self&&) s, std::move(r));
+  }
+};
+
+template <typename... Values>
+inline constexpr _connect_fn<Values...> _any_connect{};
+
+template <typename... Values>
+using _sender_base = any_unique_t<_any_connect<Values...>>;
+
+template <typename... Values>
+struct _sender {
+  struct type;
+};
+
+template <typename... Values>
+struct _sender<Values...>::type : private _sender_base<Values...> {
+  template <template <class...> class Variant, template <class...> class Tuple>
+  using value_types = Variant<Tuple<Values...>>;
+
+  template <template <class...> class Variant>
+  using error_types = Variant<std::exception_ptr>;
+
+  static constexpr bool sends_done = true;
+
+  using _sender_base<Values...>::_sender_base;
+
+  any_operation_state connect(any_receiver_of<Values...> receiver) && {
+    return _any::_any_connect<Values...>(
+        static_cast<_sender_base<Values...>&&>(*this),
+        std::move(receiver));
+  }
```

Comment:
If `receiver`'s type wasn't already erased in this context, you could know whether `set_value(receiver, values...)` is `noexcept` or not, right?  What do you think about defining `any_noexcept_receiver_of` to pair with `any_receiver_of`, and then extending `any_sender_of`'s vtable to support both `connect(any_receiver_of)` and `connect(any_noexcept_receiver_of)`?

I imagine the increased `noexcept`-ness would be a win, but it does generate a bunch of new code that couldn't be compiled out.  Could be a separate PR.

Reply:
This could be done, but it would end up double-instantiating the sender's connect() method for both the noexcept(true) and noexcept(false) cases - and then if we wanted to support multiple set_value-overloads, then for each combination of noexceptness.

If the goal of using noexcept would be to reduce binary size then I think this would probably be counter-productive.

Another approach would be to have the any_receiver_of always declare its set_value() overloads as noexcept and then if the concrete receiver's set_value is not noexcept then wrap the forwarding call to its set_value in a try/catch and call set_error() from the catch(...) clause in that case.

Reply:
> Another approach would be to have the any_receiver_of always declare its set_value() overloads as noexcept and then if the concrete receiver's set_value is not noexcept then wrap the forwarding call to its set_value in a try/catch and call set_error() from the catch(...) clause in that case.

I haven't done this. Maybe we can try it once we have `any_sender_of` in use internally and we can see what the impact is on code size.